platforms/esp/32/fastpin_esp32.h

```diff
@@ -24,68 +11,90 @@ template<uint8_t PIN, uint32_t MASK> class _ESPPIN {
   inline static void setOutput() { pinMode(PIN, OUTPUT); }
   inline static void setInput() { pinMode(PIN, INPUT); }
 
-  inline static void hi() __attribute__ ((always_inline)) { if(PIN < OUTPUT_PIN_LIMIT) { _GPB0._GPOS = MASK; } }
-  // inline static void hi() __attribute__ ((always_inline)) { gpio_set_level((gpio_num_t)PIN, HIGH); }
+  inline static void hi() __attribute__ ((always_inline)) { 
+      if (PIN < 32) GPIO.out_w1ts = MASK;
+      else if(PIN < 34) GPIO.out1_w1ts.val = MASK;
+  }
+
+  inline static void lo() __attribute__ ((always_inline)) {
+      if (PIN < 32) GPIO.out_w1tc = MASK;
+      else if(PIN < 34) GPIO.out1_w1tc.val = MASK;
+  }
 
-  inline static void lo() __attribute__ ((always_inline)) { if (PIN < OUTPUT_PIN_LIMIT){ _GPB0._GPOC = MASK; } }
-  // inline static void lo() __attribute__ ((always_inline)) { gpio_set_level((gpio_num_t)PIN, LOW); }
-  inline static void set(register port_t val) __attribute__ ((always_inline)) { if (PIN < OUTPUT_PIN_LIMIT){ _GPB0._GPO = val; }}
-  // inline static void set(register port_t val) __attribute__ ((always_inline)) { gpio_set_level((gpio_num_t)PIN, val); }
+  inline static void set(register port_t val) __attribute__ ((always_inline)) {
+      if (val) hi();
+      else lo();
+  }
 
   inline static void strobe() __attribute__ ((always_inline)) { toggle(); toggle(); }
 
-  inline static void toggle() __attribute__ ((always_inline)) { if (PIN < OUTPUT_PIN_LIMIT){ _GPB0._GPO = MASK; } }
+  inline static void toggle() __attribute__ ((always_inline)) {  }
```

Comment:
Needs an implementation - e.g. something like:

```
  GPIO_OUT_REG ^= MASK;
```

Reply:
I need to think about this one.

Reply:
This is where you would use the GPIO.out register - honestly I'm disappointed that Espressif didn't also add a toggle gpio register - but their gpio architecture seems a little hacky at times.  Anyway - this would literally just be:

```
  if(PIN < 32) { GPIO.out ^= MASK; } else { GPIO.out1.val ^=MASK; }
```